core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -952,17 +952,19 @@ class LogCleanerTest {
 
     def distinctValuesBySegment = log.logSegments.map(s => s.log.records.asScala.map(record => TestUtils.readString(record.value)).toSet.size).toSeq
 
-    val disctinctValuesBySegmentBeforeClean = distinctValuesBySegment
+    val distinctValuesBySegmentBeforeClean = distinctValuesBySegment
     assertTrue("Test is not effective unless each segment contains duplicates. Increase segment size or decrease number of keys.",
       distinctValuesBySegment.reverse.tail.forall(_ > N))
 
     cleaner.clean(LogToClean(new TopicPartition("test", 0), log, 0, firstUncleanableOffset))
 
-    val distinctValuesBySegmentAfterClean = distinctValuesBySegment
+    var distinctValuesBySegmentAfterClean = distinctValuesBySegment
 
+    assertTrue(distinctValuesBySegmentAfterClean.size == distinctValuesBySegmentBeforeClean.size - 1)
+    distinctValuesBySegmentAfterClean = 0 +: distinctValuesBySegmentAfterClean
```

Comment:
I understand the need to patch the logic here. However, the way it's done seems a bit ad hoc. Could we just change the logic to get the distinct values for a particular offset range?